File: scripts/send_email_notification.py
#!/usr/bin/env python3
"""
物件情報の変更通知メールスクリプト
"""

import json
import os
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from datetime import datetime

DATA_FILE = 'data/property_history.json'


def load_history():
    """履歴データを読み込み"""
    if os.path.exists(DATA_FILE):
        with open(DATA_FILE, 'r', encoding='utf-8') as f:
            history = json.load(f)
            return history if history else []
    return []


def load_latest_data():
    """最新のデータを読み込み"""
    history = load_history()
    if history:
        return history[-1]
    return None


def load_previous_data():
    """前回のデータを読み込み"""
    history = load_history()
    if len(history) >= 2:
        return history[-2]
    return None


def detect_changes(current_properties, previous_properties):
    """物件情報の変更を検出"""
    if not previous_properties:
        return None
    
    changes = []
    
    for current in current_properties:
        prev = next((p for p in previous_properties if p['id'] == current['id']), None)
        if not prev:
            continue
        
        # 物件数の変化
        if current.get('count', 0) != prev.get('count', 0):
            diff = current.get('count', 0) - prev.get('count', 0)
            direction = '増加' if diff > 0 else '減少'
            changes.append({
                'property': current['name'],
                'type': 'count',
                'message': f"物件数: {prev.get('count', 0)}件 → {current.get('count', 0)}件 ({abs(diff)}件{direction})"
            })
        
        # 入居時期の変化をチェック
        current_breakdown = current.get('moveInBreakdown', {})
        prev_breakdown = prev.get('moveInBreakdown', {})
        
        all_keys = set(current_breakdown.keys()) | set(prev_breakdown.keys())
        for key in all_keys:
            curr_count = current_breakdown.get(key, 0)
            prev_count = prev_breakdown.get(key, 0)
            if curr_count != prev_count:
                diff = curr_count - prev_count
                direction = '増加' if diff > 0 else '減少'
                is_march_2026 = '26年3月' in key
                changes.append({
                    'property': current['name'],
                    'type': 'move_in',
                    'key': key,
                    'is_march_2026': is_march_2026,
                    'message': f"{key}: {prev_count}件 → {curr_count}件 ({abs(diff)}件{direction})"
                })
    
    return changes if changes else None


def format_move_in_breakdown(breakdown: dict) -> str:
    """入居時期の内訳をフォーマット"""
    if not breakdown:
        return "データなし"
    
    # 26年3月を優先的に表示
    march_2026 = []
    immediate = []
    other = []
    
    for key, count in breakdown.items():
        if key == '即入居可':
            immediate.append(f"  ⚡ {key}: {count}件")
        elif '26年3月' in key:
            march_2026.append(f"  🌸 {key}: {count}件")
        else:
            other.append(f"  • {key}: {count}件")
    
    result = []
    if march_2026:
        result.append("  【26年3月入居 ★注目★】")
        result.extend(sorted(march_2026))
    if immediate:
        result.append("  【即入居可】")
        result.extend(immediate)
    if other:
        result.append("  【その他】")
        result.extend(sorted(other))
    
    return '\n'.join(result)


def count_march_2026(properties: list) -> int:
    """26年3月入居の合計を計算"""
    total = 0
    for prop in properties:
        breakdown = prop.get('moveInBreakdown', {})
        for key, count in breakdown.items():
            if '26年3月' in key:
                total += count
    return total


def create_email_content(data: dict, changes: list) -> str:
    """メール本文を作成"""
    timestamp = data.get('timestamp', '')
    date_str = data.get('date', '')
    time_str = data.get('time', '')
    properties = data.get('properties', [])
    
    # 26年3月入居の合計
    march_count = count_march_2026(properties)
    
    # 26年3月入居の変更があるかチェック
    has_march_2026_change = any(c.get('is_march_2026') for c in (changes or []))
    
    lines = [
        "=" * 50,
        "🏠 SUUMO 物件情報 変更通知",
        "=" * 50,
        f"検出日時: {date_str} {time_str}",
        "",
    ]
    
    # 変更内容を表示
    if changes:
        lines.extend([
            "📢 変更内容:",
            "-" * 40,
        ])
        
        # 物件ごとに変更をまとめる
        changes_by_property = {}
        for change in changes:
            prop_name = change['property']
            if prop_name not in changes_by_property:
                changes_by_property[prop_name] = []
            changes_by_property[prop_name].append(change)
        
        for prop_name, prop_changes in changes_by_property.items():
            lines.append(f"📍 {prop_name}")
            for change in prop_changes:
                if change.get('is_march_2026'):
                    lines.append(f"  🌸 {change['message']} ★注目★")
                else:
                    lines.append(f"  • {change['message']}")
            lines.append("")
        
        lines.append("-" * 40)
        lines.append("")
    
    # 26年3月入居があれば強調
    if march_count > 0:
        lines.extend([
            "★" * 25,
            f"🌸 26年3月入居: 現在{march_count}件",
            "★" * 25,
            "",
        ])
    
    # 各物件の情報
    total_count = 0
    for prop in properties:
        name = prop.get('name', '不明')
        count = prop.get('count', 0)
        url = prop.get('url', '')
        breakdown = prop.get('moveInBreakdown', {})
        success = prop.get('success', False)
        
        total_count += count
        
        lines.append("-" * 40)
        lines.append(f"📍 {name}")
        lines.append(f"   空室数: {count}件")
        
        if breakdown:
            lines.append("   入居時期:")
            lines.append(format_move_in_breakdown(breakdown))
        
        lines.append(f"   URL: {url}")
        lines.append("")
    
    lines.extend([
        "=" * 50,
        f"合計空室数: {total_count}件",
        "=" * 50,
        "",
        "このメールは自動送信されています。",
        "詳細はWebサイトをご確認ください。",
    ])
    
    return '\n'.join(lines)


def send_email(subject: str, body: str, to_email: str):
    """メールを送信"""
    smtp_server = os.environ.get('SMTP_SERVER', 'smtp.gmail.com')
    smtp_port = int(os.environ.get('SMTP_PORT', '587'))
    smtp_user = os.environ.get('SMTP_USER')
    smtp_password = os.environ.get('SMTP_PASSWORD')
    from_email = os.environ.get('FROM_EMAIL', smtp_user)
    
    if not smtp_user or not smtp_password:
        print("SMTP認証情報が設定されていません")
        return False
    
    msg = MIMEMultipart()
    msg['From'] = from_email
    msg['To'] = to_email
    msg['Subject'] = subject
    msg.attach(MIMEText(body, 'plain', 'utf-8'))
    
    try:
        with smtplib.SMTP(smtp_server, smtp_port) as server:
            server.starttls()
            server.login(smtp_user, smtp_password)
            server.send_message(msg)
        print(f"メール送信成功: {to_email}")
        return True
    except Exception as e:
        print(f"メール送信失敗: {e}")
        return False


def main():
    """メイン処理"""
    print(f"変更通知メール開始: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    
    # 送信先メールアドレス
    to_email = os.environ.get('NOTIFICATION_EMAIL', 'clays7614@gmail.com')
    
    # 最新データと前回データを読み込み
    data = load_latest_data()
    if not data:
        print("データがありません")
        return 1
    
    previous_data = load_previous_data()
    
    # 変更を検出
    changes = None
    if previous_data:
        changes = detect_changes(
            data.get('properties', []),
            previous_data.get('properties', [])
        )
    
    if not changes:
        print("物件情報に変更がありません")
        # 変更がなくても手動実行の場合は送信（環境変数で制御）
        if os.environ.get('FORCE_SEND') != 'true':
            return 0
        print("FORCE_SEND=true のため送信を継続します")
    
    # メール本文を作成
    body = create_email_content(data, changes)
    
    # 件名を作成
    march_count = count_march_2026(data.get('properties', []))
    date_str = data.get('date', datetime.now().strftime('%Y-%m-%d'))
    time_str = data.get('time', datetime.now().strftime('%H:%M'))
    
    # 26年3月入居の変更があるかチェック
    has_march_2026_change = any(c.get('is_march_2026') for c in (changes or []))
    
    if has_march_2026_change:
        subject = f"🌸【26年3月入居に変更あり】SUUMO物件情報 {date_str} {time_str}"
    elif march_count > 0:
        subject = f"📢【物件情報更新】SUUMO {date_str} {time_str} (26年3月: {march_count}件)"
    else:
        subject = f"📢【物件情報更新】SUUMO {date_str} {time_str}"
    
    # メール送信
    success = send_email(subject, body, to_email)
    
    print(f"変更通知メール終了: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    return 0 if success else 1


if __name__ == '__main__':
    exit(main())

File: scripts/sendEmailNotification.ts
import * as nodemailer from 'nodemailer';
import { loadHistory, HistoryEntry, PropertyData, MoveInBreakdown } from './common';

// 環境変数の定義
const SMTP_SERVER = process.env.SMTP_SERVER || 'smtp.gmail.com';
const SMTP_PORT = parseInt(process.env.SMTP_PORT || '587', 10);
const SMTP_USER = process.env.SMTP_USER;
const SMTP_PASSWORD = process.env.SMTP_PASSWORD;
const FROM_EMAIL = process.env.FROM_EMAIL || SMTP_USER;
const NOTIFICATION_EMAIL = process.env.NOTIFICATION_EMAIL || 'clays7614@gmail.com';
const FORCE_SEND = process.env.FORCE_SEND === 'true';

interface Change {
    property: string;
    type: 'count' | 'move_in';
    message: string;
    is_march_2026?: boolean;
    key?: string;
}

function loadLatestData(): HistoryEntry | null {
    const history = loadHistory();
    if (history.length > 0) {
        return history[history.length - 1];
    }
    return null;
}

function loadPreviousData(): HistoryEntry | null {
    const history = loadHistory();
    if (history.length >= 2) {
        return history[history.length - 2];
    }
    return null;
}

function detectChanges(currentProperties: PropertyData[], previousProperties: PropertyData[]): Change[] | null {
    if (!previousProperties || previousProperties.length === 0) {
        return null;
    }

    const changes: Change[] = [];

    for (const current of currentProperties) {
        const prev = previousProperties.find(p => p.id === current.id);
        if (!prev) continue;

        // 物件数の変化
        if (current.count !== prev.count) {
            const diff = current.count - prev.count;
            const direction = diff > 0 ? '増加' : '減少';
            changes.push({
                property: current.name,
                type: 'count',
                message: `物件数: ${prev.count}件 → ${current.count}件 (${Math.abs(diff)}件${direction})`
            });
        }

        // 入居時期の変化をチェック
        const currentBreakdown = current.moveInBreakdown || {};
        const prevBreakdown = prev.moveInBreakdown || {};

        const allKeys = new Set([...Object.keys(currentBreakdown), ...Object.keys(prevBreakdown)]);

        for (const key of allKeys) {
            const currCount = currentBreakdown[key] || 0;
            const prevCount = prevBreakdown[key] || 0;

            if (currCount !== prevCount) {
                const diff = currCount - prevCount;
                const direction = diff > 0 ? '増加' : '減少';
                const isMarch2026 = key.includes('26年3月');
                changes.push({
                    property: current.name,
                    type: 'move_in',
                    key: key,
                    is_march_2026: isMarch2026,
                    message: `${key}: ${prevCount}件 → ${currCount}件 (${Math.abs(diff)}件${direction})`
                });
            }
        }
    }

    return changes.length > 0 ? changes : null;
}

function formatMoveInBreakdown(breakdown: MoveInBreakdown): string {
    if (!breakdown || Object.keys(breakdown).length === 0) {
        return "データなし";
    }

    const march2026: string[] = [];
    const flexible: string[] = [];
    const scheduled: string[] = [];

    for (const [key, count] of Object.entries(breakdown)) {
        if (key.includes('26年3月')) {
            march2026.push(`  * ${key}: ${count}件`);
        } else if (['即入居可', '相談'].includes(key)) {
            flexible.push(`  * ${key}: ${count}件`);
        } else {
            scheduled.push(`  * ${key}: ${count}件`);
        }
    }

    const result: string[] = [];
    if (march2026.length > 0) {
        result.push("  【26年3月入居 *注目*】");
        march2026.sort().forEach(s => result.push(s));
    }
    if (flexible.length > 0) {
        result.push("  【即入居可・相談】");
        flexible.sort().forEach(s => result.push(s));
    }
    if (scheduled.length > 0) {
        result.push("  【時期指定】");
        scheduled.sort().forEach(s => result.push(s));
    }

    return result.join('\n');
}

function countMarch2026(properties: PropertyData[]): number {
    let total = 0;
    for (const prop of properties) {
        const breakdown = prop.moveInBreakdown || {};
        for (const [key, count] of Object.entries(breakdown)) {
            if (key.includes('26年3月')) {
                total += count;
            }
        }
    }
    return total;
}

function createEmailContent(data: HistoryEntry, changes: Change[] | null): string {
    const { timestamp, date, time, properties } = data;
    const marchCount = countMarch2026(properties);

    const lines: string[] = [
        "=".repeat(50),
        "SUUMO 物件情報 変更通知",
        "=".repeat(50),
        `検出日時: ${date} ${time}`,
        "",
    ];

    if (changes && changes.length > 0) {
        lines.push("変更内容:", "-".repeat(40));

        const changesByProperty: { [key: string]: Change[] } = {};
        for (const change of changes) {
            if (!changesByProperty[change.property]) {
                changesByProperty[change.property] = [];
            }
            changesByProperty[change.property].push(change);
        }

        for (const [propName, propChanges] of Object.entries(changesByProperty)) {
            lines.push(`[物件] ${propName}`);
            for (const change of propChanges) {
                if (change.is_march_2026) {
                    lines.push(`  * ${change.message} *注目*`);
                } else {
                    lines.push(`  - ${change.message}`);
                }
            }
            lines.push("");
        }
        lines.push("-".repeat(40));
        lines.push("");
    }

    if (marchCount > 0) {
        lines.push(
            "*".repeat(25),
            `26年3月入居: 現在${marchCount}件`,
            "*".repeat(25),
            ""
        );
    }

    let totalCount = 0;
    for (const prop of properties) {
        totalCount += prop.count;

        lines.push("-".repeat(40));
        lines.push(`[物件] ${prop.name}`);
        lines.push(`   空室数: ${prop.count}件`);
        lines.push(`   URL: ${prop.url}`);
        lines.push("");
    }

    lines.push(
        "=".repeat(50),
        `合計空室数: ${totalCount}件`,
        "=".repeat(50),
        "",
        "このメールは自動送信されています。",
        "詳細はWebサイトをご確認ください。",
        "",
        "Webサイト: https://clay7614.github.io/property_info/"
    );

    return lines.join('\n');
}

async function sendEmail(subject: string, body: string, toEmail: string): Promise<boolean> {
    if (!SMTP_USER || !SMTP_PASSWORD) {
        console.error("SMTP認証情報が設定されていません");
        return false;
    }

    const transporter = nodemailer.createTransport({
        host: SMTP_SERVER,
        port: SMTP_PORT,
        secure: false, // 587 uses STARTTLS, so secure is false
        auth: {
            user: SMTP_USER,
            pass: SMTP_PASSWORD,
        },
    });

    try {
        await transporter.sendMail({
            from: FROM_EMAIL,
            to: toEmail,
            subject: subject,
            text: body,
        });
        console.log(`メール送信成功: ${toEmail}`);
        return true;
    } catch (e) {
        console.error(`メール送信失敗: ${e}`);
        return false;
    }
}

async function main() {
    console.log(`変更通知メール開始: ${new Date().toLocaleString()}`);

    const data = loadLatestData();
    if (!data) {
        console.error("データがありません");
        process.exit(1);
    }

    const previousData = loadPreviousData();
    let changes: Change[] | null = null;

    if (previousData) {
        changes = detectChanges(data.properties, previousData.properties);
    }

    const hasMarch2026Change = changes ? changes.some(c => c.is_march_2026) : false;

    if (!changes) {
        console.log("物件情報に変更がありません");
        if (!FORCE_SEND) {
            process.exit(0);
        }
        console.log("FORCE_SEND=true のため送信を継続します");
    } else if (!hasMarch2026Change) {
        console.log("変更はありましたが、26年3月入居の物件に変化がないため送信をスキップします");
        if (!FORCE_SEND) {
            process.exit(0);
        }
        console.log("FORCE_SEND=true のため送信を継続します");
    } else {
        console.log(`26年3月入居の変更を検出しました（${changes.length}件の変更中）`);
    }

    const body = createEmailContent(data, changes);
    const marchCount = countMarch2026(data.properties);
    const dateStr = data.date;
    const timeStr = data.time;

    let subject = "";
    if (hasMarch2026Change) {
        subject = `【26年3月入居に変更あり】SUUMO物件情報 ${dateStr} ${timeStr}`;
    } else if (marchCount > 0) {
        subject = `【物件情報更新】SUUMO ${dateStr} ${timeStr} (26年3月: ${marchCount}件)`;
    } else {
        subject = `【物件情報更新】SUUMO ${dateStr} ${timeStr}`;
    }

    const success = await sendEmail(subject, body, NOTIFICATION_EMAIL);

    console.log(`変更通知メール終了: ${new Date().toLocaleString()}`);
    process.exit(success ? 0 : 1);
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});
